components/log-viewer-webui/server/src/routes/static.js

```diff
@@ -0,0 +1,55 @@
+import path from "node:path";
+import process from "node:process";
+import {fileURLToPath} from "node:url";
+
+import {fastifyStatic} from "@fastify/static";
+
+import settings from "../../settings.json" with {type: "json"};
+
+
+/**
+ * Create static files serving routes.
+ *
+ * @param {import("fastify").FastifyInstance} fastify
+ * @param {import("fastify").FastifyPluginOptions} options
+ */
+const routes = async (fastify, options) => {
+    const filename = fileURLToPath(import.meta.url);
+    const dirname = path.dirname(filename);
+    const rootDirname = path.resolve(dirname, "../..");
+
+    let irFilesDir = settings.IrFilesDir;
+    if (false === path.isAbsolute(irFilesDir)) {
+        irFilesDir = path.resolve(rootDirname, irFilesDir);
+    }
+    await fastify.register(fastifyStatic, {
+        prefix: "/ir",
+        root: irFilesDir,
+    });
+
+    let logViewerDir = settings.LogViewerDir;
+    if (false === path.isAbsolute(logViewerDir)) {
+        logViewerDir = path.resolve(rootDirname, logViewerDir);
+    }
+    await fastify.register(fastifyStatic, {
+        prefix: "/log-viewer",
+        root: logViewerDir,
+        decorateReply: false,
+    });
+
+    if ("production" === process.env.NODE_ENV) {
+        // In the development environment, we expect the client to use a separate webserver that
+        // supports live reloading.
+        if (false === path.isAbsolute(settings.ClientDir)) {
+            throw new Error("`clientDir` must be an absolute path.");
+        }
+
+        await fastify.register(fastifyStatic, {
+            prefix: "/",
+            root: settings.ClientDir,
```

Comment:
What if path itself is an empty string? This resolve function will return false, will that cause a problem?

Reply:
> What if path itself is an empty string? 

This is some good thinking. Previously if we pass an empty string directly to the `root` field, an error would be reported by fastifyStatic saying the path isn't an absolute one, and the user would be notified. 

> This resolve function will return false

That doesn't seem to be the case on my end. `path.resolve("/absolute/path/to/clp/components/log-viewer-webui/server", "")` returns `/absolute/path/to/clp/components/log-viewer-webui/server` which is a valid absolute path. I think it is okay for us to allow the user to set an empty string for the paths to reference the server project root (although practically that's not good for security reasons and even if so users should have put "." instead of an empty string).